cmd/kops/get_keypairs.go

```diff
@@ -122,30 +122,39 @@ func listKeypairs(keyStore fi.CAStore, names []string, includeDistrusted bool) (
 		}
 
 		for _, item := range keyset.Items {
-			if includeDistrusted || item.DistrustTimestamp == nil {
-				var alternateNames []string
-				alternateNames = append(alternateNames, item.Certificate.Certificate.DNSNames...)
-				alternateNames = append(alternateNames, item.Certificate.Certificate.EmailAddresses...)
-				for _, ip := range item.Certificate.Certificate.IPAddresses {
-					alternateNames = append(alternateNames, ip.String())
-				}
-				sort.Strings(alternateNames)
-
+			if includeDistrusted || (item.DistrustTimestamp == nil && item.Certificate != nil) {
```

Comment:
It's arguably strange to overload the meaning of "distrusted" here. We could update the documentation to mention this additional consideration, or introduce a separate flag to include these items lacking a certificate. I'm not sure that it's worth it.

Reply:
I think `ToCertificateBytes()` and `ToPublicKeys()` in `fi.Keyset` need to also have `Certificate != nil` checks in addition to their current `DistrustTimestamp` checks, otherwise they could also trigger nil dereferences. 

Given the above were done, it would be safe to claim that keypairs without certificates are inherently distrusted.

Similarly, `VFSCAStore.StoreKeyset` should verify that `Certificate != nil` for the primary id. promote_keypair.go should refuse to promote a keypair without a certificate.

Reply:
John, I noticed that [`(*fi.Keyset).ToCertificateBytes` filters items that lack a certificate](https://sourcegraph.com/github.com/kubernetes/kops@92e856e5016ce295d414901f85efa6a960f6e724/-/blob/upup/pkg/fi/ca.go?L150-L156). It could filter them a little bit earlier, but it looks they already won't make it into the composed CA bundle. Likewise, [`(*fi.Keyset).ToPublicKeys` is guarded in the same way](https://sourcegraph.com/github.com/kubernetes/kops@92e856e5016ce295d414901f85efa6a960f6e724/-/blob/upup/pkg/fi/ca.go?L175-L183).

Reply:
Okay, so it seems promote_keypair.go is the only remaining issue.

Reply:
I _did_ augment `(*fi.VFSCAStore).StoreKeyset`. I'm now working on the keypair promotion.

Reply:
Please take another look, John. Thank you for the advice so far.